text/0152-telemetry-schemas.md

```diff
@@ -0,0 +1,1159 @@
+proposal).
+
+All schemas therefore form a set of rooted trees, where nodes represent schema
+versions and pairs of nodes are connected via edges of two types: 1)
+parent-child relationship, 2) consecutive versions in the same schema family.
+Each edge in the tree represents a schema transformation.
+
+Once we have the tree of schemas it is possible to then convert from a node in a
+tree to another node in a tree given that the nodes are connected via a path
+through edges.
+
+Further research is needed in this area if we find that we need this concept of
+parent schemas.
+
+### Current State in Schema
+
+The schema file in this OTEP describes changes from version to version. This is
+very useful for performing conversions between versions but it does not capture
+the current state of a particular version.
+
+We can add an ability to specify the full current schema of each particular
+version in the schema file (in addition to the "changes" section). This will
+open up a few interesting possibilities:
+
+- Automatic validation that emitted telemetry confirms to the declared schema.
+
+- OpenTelemetry semantic conventions document can be automatically generated
+  from this formal schema file. This will remove the need to have semantic
+  conventions codified and generated from yaml files in the specification.
+
+- Consumers of telemetry can use this information to interpret the received
+  telemetry.
+
+- It may be possible to auto-generate the "changes" section in schema files as a
+  delta between full states of versions.
+
+An example of what can be recorded as the "current state" is the encoding of the
+Body of the Log Records. It was described as a concept in
+[this earlier proposal](https://docs.google.com/document/d/1ZExye1lW43owwaxcbjOvl0P2qER-UaYd_MyxItc2h0k/edit#).
+The suggestion was to record the encoding of the Body field of Log Records,
+which can be also implemented as a new setting in the "current state" of the
+"logs" section in the schema file.
+
+### Other Transformation Types
+
+This OTEP introduces a limited set of transformations, while deliberately
+keeping the type of transformations to a minimum. It is easy to see that
+telemetry can evolve in other, more sophisticated ways over time (see e.g.
+[this list](https://github.com/open-telemetry/opentelemetry-specification/issues/1324)).
+
+More transformation types can be added in the future when there is a need to
+represent a particular change in the telemetry schema that is not possible with
+the current set of transformations. However, care should be taken to avoid
+introducing unnecessary or overly complicated transformation types, especially
+ones that cannot be applied on a local portion of telemetry data and requires a
+full state (e.g. transformations, such as aggregations, etc), since such
+transformations can place significant implementation burden on telemetry sources
+that wish to support the notion of telemetry schemas.
+
+## Alternates Considered
```

Comment:
This is a rather complex and powerful approach. To justify it, I think there is one more alternative to consider (with corresponding subsection here): Have a strict policy to never rename an attribute once it becomes stable. While this may lead to sub-optimal names, it would be much simpler to implement and have not negative performance impact.

Reply:
This is surely up for debate.
I think I have heard from enough sources that fossilizing semantic conventions and emitted telemetry schema is not an option. IMO, long term it is an impossible position. People make mistakes and want to be able to fix them. The eventual surface area of semantic convention and telemetry schemas as a whole is much bigger than the API.
Forcedly fixing the API is achievable (especially given that we have tools like compilers that help enforce it). Fixing the semantic convention I think is much more difficult (especially given that there is no fully machine-readable definition of Otel schema and there are no tools that can help enforce that schema). We could introduce such machine-readable schema and tools, but the effort for Otel will be bigger than what this proposal asks for and still won't be as useful for enabling the evolution of schemas.

Reply:
That sounds reasonable, but I would add a subsection on this to the "Alternates Considered".

Reply:
Sounds good, I will add.

Reply:
Added a subsection "Freeze Schema" in "Alternates Considered".